File: libSOS/server/src/robotmessagequeue/robot_command.cpp
/*
 * robot_command.h
 *
 *  Created on: Nov 1, 2013
 *      Author: jamie
 */

#include "robot_command.h"
#include <vector>
#include <boost/foreach.hpp>
#include "../libSOS/SOSProtocol.h"
#include <sstream>
#include <LogMacros.h>
#include <iomanip>

std::string string_to_hex(const std::string& input)
{
    static const char* const lut = "0123456789ABCDEF";
    size_t len = input.length();

    std::string output;
    output.reserve(2 * len);
    for (size_t i = 0; i < len; ++i)
    {
        const unsigned char c = input[i];
        output.push_back(lut[c >> 4]);
        output.push_back(lut[c & 15]);
    }
    return output;
}

robot_command::robot_command(in_port_t return_id, unsigned char opcode)
:_return_id(return_id),
_opcode(opcode),
 _params(),
 _extraString()
{

}

robot_command::robot_command(in_port_t return_id, unsigned char opcode, boost::optional<std::vector<param_t> > data)
:_return_id(return_id),
_opcode(opcode),
 _params(data),
 _extraString()
{

}

robot_command::robot_command(in_port_t return_id, unsigned char opcode, boost::optional<std::vector<param_t> > data, boost::optional<std::string> extraString)
:_return_id(return_id),
_opcode(opcode),
 _params(data),
 _extraString(extraString)
{

}

boost::optional<robot_command> robot_command::factory(std::vector<char> bytes)
{
	std::vector<char>::const_iterator 		currentBytePtr(bytes.begin());
	in_port_t								return_id;
	char 									opcode;
	boost::optional<std::vector<param_t> > 	params;
	boost::optional<std::string> 			string;


	return_id = parse_return_id(currentBytePtr, bytes);

	++currentBytePtr;
	opcode = parse_opcode(currentBytePtr, bytes);

	// Move into params if any.
	++currentBytePtr;
	// should NOW equal END_TRANSMISSION or START_SHORT_TRANSMISSION

	std::cout << "command: " << std::setw(2) << string_to_hex(std::string(bytes.begin(), bytes.end())) << std::endl;

	//returns an empty optional if there isn't a param
	params = parse_params(currentBytePtr, bytes);

	string = parse_string(currentBytePtr, bytes);

	return boost::optional<robot_command>(robot_command(return_id, opcode, params, string));

}

#if 0

	std::vector<char> myVec(currentBytePtr, std::find(currentBytePtr, endBytePtr, END_ID));

#endif

in_port_t robot_command::parse_return_id(std::vector<char>::const_iterator & currentBytPtr, std::vector<char> & bytes)
{
	std::vector<char>::const_iterator currentBytePtr = std::find(bytes.cbegin(), bytes.cend(), START_TRANSMISSION);
	if(currentBytePtr == bytes.end())
	{
		// Bail and return empty value.
		std::cerr << "RobotCommand: the header of the command is incorrect\nShould be:" << START_TRANSMISSION << " Was: " << *currentBytePtr << std::endl;
		return 0x0;
	}

	// Have start of command.  Begin parsing.


	++currentBytePtr;
	//should now be the first byte of the ASCII-encoded id

	++currentBytePtr;

	//copy elements from currentBytePtr to END_ID into IDStorage
	std::vector<char>::const_iterator endIDPtr=std::find(currentBytePtr, bytes.cend(), (char)(END_ID));
	std::vector<char> IDStorage(currentBytePtr, (endIDPtr));

	//backstop atoi
	IDStorage.push_back(0x0);

	//move iterator to IDStorage's position
	//currentBytePtr += std::distance(currentBytePtr, endIDPtr);

	//should now equal END_ID
	return static_cast<in_port_t>(atoi(IDStorage.data()));
}

char robot_command::parse_opcode(std::vector<char>::const_iterator & currentBytPtr, std::vector<char> & bytes)
{
	std::vector<char>::const_iterator currentBytePtr = std::find(bytes.cbegin(), bytes.cend(), START_OPCODE);

	// Get opcode.
	++currentBytePtr;
	char opcode = *currentBytePtr;
	++currentBytePtr;

	if(*currentBytePtr != END_OPCODE)
	{
		// Bail and return empty value.
		LOG_RECOVERABLE("RobotCommand", " the end of the opcode command is incorrect\nShould be:" << END_OPCODE << " Was: " << *currentBytePtr);
		return 0x0;
	}
	return opcode;
}

boost::optional<std::vector<robot_command::param_t> > robot_command::parse_params(std::vector<char>::const_iterator & currentBytPtr, std::vector<char> & bytes)
{
	std::vector<char>::const_iterator currentBytePtr = std::find(bytes.cbegin(), bytes.cend(), START_PARAMS);

	//create an empty optional
	boost::optional<std::vector<param_t> > params;


	// Handle case where START_SHORT_TRANSMISSION does not occur.
	if(currentBytePtr == bytes.cend())
	{
		return params;
	}


	//give the optional type a value
	params.reset(std::vector<robot_command::param_t>());


	 // Loop until we don't recieve the start of a param
	while(*currentBytePtr == START_PARAMS)
	{
		//copy elements from currentBytePtr to END_PARAMS into paramBytesStorage
		std::vector<char>::const_iterator endShortPtr = std::find(currentBytePtr, bytes.cend(), (char)(END_PARAMS));
		std::vector<char> paramBytesStorage((currentBytePtr + 1), endShortPtr);

		//backstop atoi
		paramBytesStorage.push_back(0x0);

		currentBytePtr += std::distance(currentBytePtr, endShortPtr);

		params.get().push_back(strtod(paramBytesStorage.data(), NULL));

		++currentBytePtr;
	}

	return params;

}


boost::optional<std::string> robot_command::parse_string(std::vector<char>::const_iterator & currentBytPtr, std::vector<char> & bytes)
{
	std::vector<char>::const_iterator currentBytePtr = std::find(bytes.cbegin(), bytes.cend(), START_STRING);

	boost::optional<std::string> optional_string;

	if(currentBytePtr == bytes.cend())
	{
		return optional_string;

	}

	++currentBytePtr;

	std::vector<char>::const_iterator endStringPtr=std::find(currentBytePtr, bytes.cend(), (char)(END_STRING));
	std::string string(currentBytePtr, endStringPtr);

	optional_string.reset(string);
	return optional_string;

}


std::ostream & operator<<(std::ostream & leftOp, const robot_command rightOp)
{
	leftOp << "robot_command dump: --------------------------------------" << std::endl;

	leftOp << "Return ID: " << std::dec << rightOp._return_id << std::endl;

	leftOp << "Opcode: 0x" << std::hex << ((int)rightOp._opcode) << std::endl;
	if(rightOp._params.is_initialized())
	{
		leftOp << "Data: ";
		for(robot_command::param_t subdata : rightOp._params.get())
		{
			leftOp << std::dec << subdata << " ";
		}
		leftOp << std::endl;
	}
	if(rightOp._extraString.is_initialized())
	{
		leftOp << "Extra String: \"" << rightOp._extraString.get() << "\"" << std::endl;
	}
	leftOp << std::endl;

	return leftOp;

}

File: FRC-3128-2014/src/frc3128/Util/Connection/RobotCommand.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package frc3128.Util.Connection;

import frc3128.Util.DebugLog;
import java.util.Vector;

/**
 *
 * @author jamie
 */
public class RobotCommand {

    public int _return_id;
    public byte _opcode;
    public Vector _params;
    public String _string;

    public RobotCommand(int return_id, byte opcode) {
        _return_id = return_id;
        _opcode = opcode;
    }

    public RobotCommand(int return_id, byte opcode, Vector shorts) {
        _return_id = return_id;
        _opcode = opcode;
        _params = shorts;
    }

    public RobotCommand(int return_id, byte opcode, Vector shorts, String extraString) {
        _return_id = return_id;
        _opcode = opcode;
        _params = shorts;
        _string = extraString;
    }

    public RobotCommand(int return_id, byte opcode, String extraString) {
        _opcode = opcode;
        _string = extraString;
    }

    public String toString() {
        String string = "Robot Command dump:---------------\n";
        string += "Return ID: " + _return_id + "\n";
        string += "Opcode: " + Integer.toHexString(_opcode) + "\n";
        //Print out each short
        if (_params != null) {
            string += "Shorts: ";
            for (int counter = 0; counter < _params.size(); ++counter) {
                string += _params.elementAt(counter) + " ";
            }
            string += "\n";
        }

        if (_string != null) {
            string += "Extra String: " + _string + "\n";
        }
        return string;
    }

    //Parsing code originally written in c++, where it was a lot... better
    public static RobotCommand factory(byte[] currentBytePtr) {
        int return_id = parseReturnID(currentBytePtr);
        ++iterator;
        byte opcode = parseOpcode(currentBytePtr);
        iterator++;
        Vector shorts = parseParams(currentBytePtr);
        String extraString = parseString(currentBytePtr);
        iterator = 0;
        return new RobotCommand(return_id, opcode, shorts, extraString);
    }
    //needs to be class scope so that it can keep its state after a function call
    static int iterator = 0;

    static byte parseOpcode(byte[] currentBytePtr) {
        // Get opcode.
        ++iterator;
        byte opcode = currentBytePtr[iterator];
        ++iterator;

        if (currentBytePtr[iterator] != SOSProtocol.END_OPCODE) {
            // Bail and return empty value.
            DebugLog.log(DebugLog.LVL_ERROR, null, "RobotCommand: the end of the opcode command is incorrect\nShould be:" + SOSProtocol.END_OPCODE + " Was: " + currentBytePtr[iterator]);
            return 0x0;
        }

        return opcode;
    }

    static int parseReturnID(byte[] currentBytePtr) {
        if (currentBytePtr[iterator] != SOSProtocol.START_TRANSMISSION) {
            // Bail and return empty value.
            DebugLog.log(DebugLog.LVL_ERROR, "RobotCommand", "the header of the command is incorrect\nShould be:" + SOSProtocol.START_TRANSMISSION + " Was: " + currentBytePtr[iterator]);
            return 0x0;
        }
        // Have start of command.  Begin parsing.
        ++iterator;
        //should now be the first byte of the ASCII-encoded id

        //max of 10 ASCII digit short
        char[] returnBytesStorage = new char[10];
        int returnBytesIterator = 0;

        while (currentBytePtr[++iterator] != SOSProtocol.END_ID) {
            returnBytesStorage[returnBytesIterator] = (char) currentBytePtr[iterator];
            ++returnBytesIterator;
        }
        return Integer.parseInt(String.valueOf(returnBytesStorage, 0, returnBytesIterator));
    }

    static Vector parseParams(byte[] currentBytePtr) {
        if (currentBytePtr[iterator] == SOSProtocol.START_PARAM) {
            Vector shortVector = new Vector();

            // Loop until we don't recieve the start of a short
            while (currentBytePtr[iterator] == SOSProtocol.START_PARAM) {
                //max of 10 ASCII digit short
                char[] paramBytesStorage = new char[10];
                int paramBytesIterator = 0;
                ++iterator;
                while (currentBytePtr[iterator] != SOSProtocol.END_PARAM) {
                    paramBytesStorage[paramBytesIterator++] = (char) currentBytePtr[iterator++];
                }
                shortVector.addElement(Double.valueOf(String.valueOf(paramBytesStorage, 0, paramBytesIterator)));
                ++iterator;
            }
            return shortVector;
        } else {
            return null;
        }

    }

    static String parseString(byte[] currentBytePtr) {
        if (currentBytePtr[iterator] == SOSProtocol.START_STRING) {
            String stringstream = new String();
            while (currentBytePtr[++iterator] != SOSProtocol.END_STRING) {
                stringstream += currentBytePtr[iterator];
            }
            return stringstream;
        }
        return null;
    }
    
    public String params() {
        return (this._params == null ? "" : this._params.toString());
    }
/**
 * Encodes this RobotCCommand encodes it back into a binary transmission
 * @param command
 * @return 
 */
    byte[] reencodeCommand()
    {
        byte[] returnIDBytes = reencodeReturnID();
        
        byte[] opcodeBytes = reencodeOpcode();
        
        byte[] paramBytes = null;
        
        if(_params != null)
        {
            paramBytes = reencodeParams();
        }
        
        byte[] stringBytes = {0};
        
        if(_string != null) 
        {
            stringBytes = reencodeString();
        }
        
        //create the final array, adding up the total lengths of all the parts if they exist
        byte[] commandBytes = new byte
        [
                returnIDBytes.length + 
                opcodeBytes.length + 
                (_params != null ? paramBytes.length : 0) + 
                (_string != null ? stringBytes.length : 0) +
                1 /* for the end transmission character*/
        ];
        
        System.arraycopy(returnIDBytes, 0, commandBytes, 0, returnIDBytes.length);
        System.arraycopy(opcodeBytes, 0, commandBytes, returnIDBytes.length, opcodeBytes.length);
        
        if(_params != null)
        {
            System.arraycopy(paramBytes, 0, commandBytes, (returnIDBytes.length + opcodeBytes.length), paramBytes.length);
        }
        
        if(_string != null)
        {
            System.arraycopy(stringBytes, 0, commandBytes, (returnIDBytes.length + opcodeBytes.length + paramBytes.length), stringBytes.length);
        }
        
        //add end transmission character
        commandBytes[commandBytes.length - 1] = SOSProtocol.END_TRANSMISSION;
        
        return commandBytes;
    }
    
    private byte[] reencodeReturnID() 
    {
        String returnID = SOSProtocol.START_TRANSMISSION + 
                SOSProtocol.START_ID +
                Integer.toString(_return_id) +
                SOSProtocol.END_ID;
        
        return returnID.getBytes();
    }
    
    private byte[] reencodeOpcode() 
    {
        byte[] opcodeBytes = new byte[]{SOSProtocol.START_OPCODE, _opcode, SOSProtocol.END_OPCODE};
        
        return opcodeBytes;
    }
    
    private byte[] reencodeParams() 
    {
        String paramBytes = Integer.toString(SOSProtocol.START_PARAM);
        
        for(int counter = 0; counter < _params.size(); --counter)
        {
            paramBytes += _params.elementAt(counter).toString();
        }
        
        paramBytes += SOSProtocol.END_PARAM;
        
        return paramBytes.getBytes();
    }
    
    private byte[] reencodeString() 
    {
        String string = SOSProtocol.START_STRING + 
                _string +
                SOSProtocol.END_STRING;
        
        return string.getBytes();
    }


}